d2/src/test/java/com/linkedin/d2/balancer/strategies/framework/LoadBalancerStrategyTestRunnerBuilder.java

```diff
@@ -0,0 +1,494 @@
+        .setDegraderStrategies(new HashMap<>(), DEGRADER_PROPERTIES_WITH_HIGH_LOW_ERROR)
+        .build();
+  }
+
+
+  /**
+   * Mock a transport client, the transport client leverages the clockedExecutor to control the latency
+   */
+  class MockTransportClient implements TransportClient
+  {
+    private final ClockedExecutor _clockedExecutor;
+    private final LatencyManager _latencyManager;
+    private final ErrorCountManager _errorCountManager;
+    private final URI _uri;
+    private final long _intervalMillis;
+    private final Object _lock;
+
+    private Map<URI, Integer> _currentErrorCountMap;
+    private Map<URI, Integer> _lastRequestCountMap;
+    private Map<URI, Integer> _callCountMap;
+    private Map<URI, Long> _latencySumMap;
+
+    MockTransportClient(
+        ClockedExecutor executor, LatencyManager latencyManager, ErrorCountManager errorCountManager, URI uri,
+        long intervalMillis, Object lock, Map<URI, Integer> currentErrorCountMap, Map<URI, Integer> lastRequestCountMap,
+        Map<URI, Integer> callCountMap, Map<URI, Long> latencySumMap)
+    {
+      _clockedExecutor = executor;
+      _latencyManager = latencyManager;
+      _errorCountManager = errorCountManager;
+      _uri = uri;
+      _intervalMillis = intervalMillis;
+
+      _lock = lock;
+      _currentErrorCountMap = currentErrorCountMap;
+      _lastRequestCountMap = lastRequestCountMap;
+      _callCountMap = callCountMap;
+      _latencySumMap = latencySumMap;
+    }
+
+    @Override
+    public void restRequest(RestRequest request, RequestContext requestContext, Map<String, String> wireAttrs,
+        TransportCallback<RestResponse> callback)
+    {
+      int currentIntervalIndex = (int) (_clockedExecutor.currentTimeMillis() / _intervalMillis);
+      int requestCount = _lastRequestCountMap.getOrDefault(_uri, 0);
+      long latency = _latencyManager.getLatency(_uri, requestCount, currentIntervalIndex);
+      boolean hasError = _errorCountManager.getErrorCount(_uri, requestCount, currentIntervalIndex) -
+          _currentErrorCountMap.getOrDefault(_uri, 0) > 0;
+
+      _clockedExecutor.schedule(new Runnable()
+      {
+        @Override
+        public void run()
+        {
+          RestResponseBuilder restResponseBuilder = new RestResponseBuilder().setEntity(request.getURI().getRawPath().getBytes());
+          if (hasError)
+          {
+            restResponseBuilder.setStatus(HttpStatus.INTERNAL_SERVER_ERROR_500);
+            RestException restException = new RestException(restResponseBuilder.build(), new Throwable("internal error"));
+            callback.onResponse(TransportResponseImpl.error(restException));
+            return;
+          }
+          callback.onResponse(TransportResponseImpl.success(restResponseBuilder.build()));
+        }
+      }, latency, TimeUnit.MILLISECONDS);
+
+      // Collect basic stats
+      synchronized (_lock)
```

Comment:
Will mockTransportClient handle concurrent requests? I'm trying to understand why we need to have a lock.

Reply:
You're right, there is only 1 thread calling the tranportClient.restRequest sequentially, I don't need this lock, just removed it.